### PR TITLE
Make replacement behavior for HasMany relationship configurable

### PR DESCRIPTION
**Pull request by X-Coder264**
Closes https://github.com/laravel-json-api/laravel/issues/73

### CODE REVIEW

**Review by lindyhopchris — CHANGES_REQUESTED** [review]
For the tests, you should still use the comments relationship on the `users` resource. All you need to do in the tests that assert the related models are deleted is modify the field to mark it as deleted detached records.

To do that, update the has-many `TestCase` to this:

```php

    /**
     * @var UserSchema
     */
    protected UserSchema $schema;

    /**
     * @var Repository
     */
    protected Repository $repository;

    /**
     * @inheritDoc
     */
    protected function setUp(): void
    {
        parent::setUp();

        $this->schema = $this->schemas()->schemaFor('users');
        $this->repository = $this->schema->repository();
    }
```

Then at the start of each test where you are checking it deletes the related records, all you need to do right at the start of test is this:

```php
$this->schema->relationship('comments')->deleteDetachedModels();
```

Hopefully that makes sense!

File: src/Fields/Relations/HasMany.php [hunk 1]
```diff
@@ -33,6 +33,16 @@ class HasMany extends ToMany implements FillableToMany
...
 
+    private const DETACH = 'detach';
+    private const DELETE = 'delete';
+
+    /**
+     * Whether the relationship is gonna be detached or deleted when a replacement request no longer has the model.
+     *
+     * @var string
+     */
+    private $replacementBehaviorForNoLongerExistingResources = self::DETACH;
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply src/Fields/Relations/HasMany.php]: :+1: Done.

File: src/Fields/Relations/HasMany.php [hunk 2]
```diff
@@ -45,6 +55,30 @@ public static function make(string $fieldName, string $relation = null): HasMany
         return new self($fieldName, $relation);
     }
 
+    /**
+     * The relationship is gonna be detached when a replacement request no longer has the model.
+     *
+     * @return $this
+     */
+    public function detachNoLongerExistingRecordsOnReplacementRequest(): self
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply src/Fields/Relations/HasMany.php]: Done.

File: src/Fields/Relations/HasMany.php [hunk 3]
```diff
@@ -45,6 +55,30 @@ public static function make(string $fieldName, string $relation = null): HasMany
...
+
+        return $this;
+    }
+
+    /**
+     * The relationship is gonna be deleted when a replacement request no longer has the model.
+     *
+     * @return $this
+     */
+    public function deleteNoLongerExistingRecordsOnReplacementRequest(): self
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply src/Fields/Relations/HasMany.php]: Done.

File: src/Fields/Relations/HasMany.php [hunk 4]
```diff
@@ -123,7 +157,11 @@ private function doDetach(Model $model, EloquentCollection $remove): void
                 $model->setAttribute($relation->getMorphType(), null);
             }
 
-            $model->setAttribute($relation->getForeignKeyName(), null)->save();
+            if (self::DELETE === $this->replacementBehaviorForNoLongerExistingResources) {
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply src/Fields/Relations/HasMany.php]: Done.

File: tests/lib/Acceptance/Relations/HasMany/DetachTest.php [hunk 5]
```diff
@@ -73,7 +74,50 @@ public function test(): void
...
+         */
+        $this->assertFalse($user->relationLoaded('posts'));
+
+        $this->assertDatabaseHas('posts', [
+            'id' => $keep->getKey(),
+            'user_id' => $user->getKey(),
+        ]);
+
+        foreach ($remove as $post) {
+            $this->assertNull(Post::find($post->getKey()));
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply tests/lib/Acceptance/Relations/HasMany/DetachTest.php]: Done (the only reason it was written like this is that the post model is soft deletable so the assertDatabaseMissing approach was not working for it, there is no issue now that the comments relationship is being used to test this feature).

File: tests/lib/Acceptance/Relations/HasMany/TestCase.php [hunk 6]
```diff
@@ -46,10 +46,10 @@ protected function setUp(): void
      * @param iterable $actual
      * @return void
      */
-    protected function assertComments(iterable $expected, iterable $actual): void
+    protected function assertModels(iterable $expected, iterable $actual): void
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply tests/lib/Acceptance/Relations/HasMany/TestCase.php]: Done.

File: tests/lib/Acceptance/Relations/HasMany/SyncTest.php [hunk 7]
```diff
@@ -70,6 +71,46 @@ public function test(): void
         ]);
     }
 
+    public function testSyncDeletesDeletableRelationship(): void
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply tests/lib/Acceptance/Relations/HasMany/SyncTest.php]: Done.

File: tests/lib/Acceptance/Relations/HasMany/SyncTest.php [hunk 8]
```diff
@@ -70,6 +71,46 @@ public function test(): void
...
+        $this->assertSame($actual, $user->getRelation('posts'));
+
+        foreach ($expected as $post) {
+            $this->assertDatabaseHas('posts', [
+                'id' => $post->getKey(),
+                'user_id' => $user->getKey(),
+            ]);
+        }
+
+        $this->assertNull(Post::find($remove->getKey()));
```
[the remark this answers is not in the document]
Comment (X-Coder264) [reply tests/lib/Acceptance/Relations/HasMany/SyncTest.php]: Done.

**Pull request merged**